Add flags on logs

seguins opened the pull request:
This PR adds the --follow, --timestamps and --tail options on logs (#2227).

dnephin commented on compose/cli/log_printer.py:
```diff
@@ -13,11 +13,21 @@
+                 cascade_stop=False,
+                 follow=False,
+                 timestamps=False,
+                 tail="all"):
```
Instead of passing these as individual properties, could we do it as a single dict `log_args` ?

dnephin commented:
Thanks for this PR! This is looking good, although I'm not sure if we can just change the api call from attach to logs. We'll have to make sure we have reasonable test coverage in this area, and try it out manually.

The test failure is unrelated I think.

seguins commented:
I transformed the logs args as a dictionary.

I added some tests, mainly acceptance. I'm sure you would like integration or units tests, but I don't know how to do that because it's mainly api calls ?

j-san commented:
+1 for tail

dnephin commented on compose/cli/main.py:
```diff
@@ -308,13 +308,24 @@ def logs(self, project, options):
+        tail = options['--tail']
+        if tail is not None and tail.isdigit():
+            tail = int(tail)
+        log_args['tail'] = tail
```
Small problem here, we should show an error if `--tail` is invalid instead of just ignoring it.   

`image_type_from_opt()` is a good example for this. It lets you move this `if` branch into the validation function as well.

dnephin commented on compose/utils.py:
```diff
@@ -49,6 +49,9 @@ def split_buffer(stream, splitter=None, decoder=lambda a: a):
     buffered = six.text_type('')
 
+    if type(stream) is bytes:
+        stream = stream.decode("utf-8")
```
This should already be handled by `stream_as_text()`.

This also seems wrong because string should never be bytes, it should be an iterator.

seguins replied:
In python3, with option stream as False, `container.logs` returns an object of type `<class 'bytes'>`, in python2, the same api returns `<type 'str'>`. I converted bytes to string to patch it, but maybe is a bug of dockerpy ?

dnephin replied:
I see what's happening. I missed it before but setting `stream=False` is actually wrong.

I think we'll want to remove this, and I'll add another comment for the other fix.

dnephin commented on tests/fixtures/logs-composefile/docker-compose.yml:
```diff
@@ -0,0 +1,6 @@
+  command: sh -c "echo hello && sleep 5 && echo world"
+another:
+  image: busybox:latest
+  command: sh -c "sleep 6 && echo test"
```
These sleeps make the tests excessively slow. It also makes them potentially flaky if running on a heavily loaded server.

I believe the `sleep` is used to test follow vs no-follow behaviour.
What do you think about doing this instead:
* have `simple` service `sleep 200` , and `another` service without sleep
* when testing follow you can assert that the process is still running after getting the first log line out. Then `proc.terminate()` (or `os.kill()`) the logs command so we don't wait on the follow to finish. You can do this by calling `start_process()` instead of dispatch (there are a few examples of this).

Testing no-follow would be the same.
Testing other flags should never use `-f` so that we don't wait for logs to finish.

dnephin commented:
Thanks! This is looking great.  Couple comments around error reporting and testing.

dnephin commented on compose/cli/main.py:
```diff
@@ -308,13 +308,24 @@ def logs(self, project, options):
 
         monochrome = options['--no-color']
+        log_args = {}
+        log_args['stream'] = options['--follow']
```
`stream` and `follow` should be different things, however it looks like they were set to the same thing in `docker-py`. `stream` should say "return an iterator", where as follow is a behaviour of the API to continue to send logs or stop once it receives the latest one.

I think we need to fix `docker-py` to have separate params, then this should be:

```
logs_args['follow'] = ...
```

`stream` should always be `True`.

seguins replied:
I understand the difference between `stream` and `follow`, but I'm not sure to understand the advantage of this change.

If you set `follow` to `True`, should `stream` always be `True` ?

Maybe should we open an issue on `docker-py` to talk about it before to merge this PR ?

dnephin replied:
Yes, I think `follow=True` does imply `stream=True`, but `stream=True` doesn't imply `follow=True`, you may still want to stream without following.

Yes, I created #934

seguins commented:
Thanks for the feedback, I added an error message when the tail parameter is invalid, and I rewrote some tests.

I'm waiting for the issue docker/docker-py#934 to finish this PR.

SvenDowideit commented:
the documentation checker failure is due to the tutorials repo going private

https://github.com/docker/compose/pull/2928 will solve that when merged.

dnephin commented:
Cool, the docker-py change has been merged and will be in the next release.

For now you can use a git url and git sha for the docker-py dependency in `requirements.txt`: For example: `git+https://github.com/docker/docker-py.git@81d8caaf36159bf1accd86eab2e157bf8dd071a9#egg=docker-py`

seguins commented:
I changed the docker-py dependency. Now, we always call the `logs` api with `stream=True`.

dnephin commented on compose/cli/log_printer.py:
```diff
@@ -13,11 +13,18 @@
+                 output=sys.stdout,
+                 monochrome=False,
+                 cascade_stop=False,
+                 log_args={'follow': False}):
```
Using a mutable value as a default kwarg value can lead to a lot of confusion.

Instead of this default, please use

```python

def __init__(self, ..., log_args=None):
    log_args = log_args or {}
    ...
```

dnephin commented on compose/cli/log_printer.py:
```diff
@@ -64,28 +71,30 @@ def get_log_generator(container):
         prefix,
         container.log_driver)
-    yield color_func(wait_on_exit(container))
+    if log_args['follow']:
```
I think this should be `log_args.get('follow')` in case the follow key doesn't exist.

dnephin commented on compose/cli/log_printer.py:
```diff
@@ -64,28 +71,30 @@ def get_log_generator(container):
     for line in line_generator:
         yield prefix + line
-    yield color_func(wait_on_exit(container))
+    if log_args['follow']:
```
same here

dnephin commented on tests/acceptance/cli_test.py:
```diff
@@ -94,6 +95,32 @@ def __str__(self):
+        self.proc = proc
+        self.lines = ''
+        t = threading.Thread(target=self.read_thread)
+        t.start()
```
You shouldn't need any threads here. The Compose binary is run in a subprocess, which can be invoked without blocking.

dnephin commented on tests/acceptance/cli_test.py:
```diff
@@ -1141,6 +1168,47 @@ def test_unpause_no_containers(self):
+        assert 'hello' in result
+        assert 'test' in result
+        service = self.project.get_service('simple')
+        self.assertEqual(len(service.containers()), 1)
```
Minor: we've been moving away from these `unittest` assertions, to use bare `assert` statements instead, which work much better with `pytest`. It would be good to just make this `assert len(...) == 1`

dnephin commented:
This is looking good. I'm a little surprised there isn't any changes to `def up()`. I suspect that it's going to need to pass in `follow=True`, and it's a bit concerning we don't have any tests failing without it.

aanand commented on compose/cli/main.py:
```diff
@@ -328,13 +328,27 @@ def logs(self, project, options):
         monochrome = options['--no-color']
+        log_args = {}
+        log_args['follow'] = options['--follow']
+        log_args['timestamps'] = options['--timestamps']
```
Minor thing, but this could be a single assignment with a dictionary literal.

```python
log_args = {
    'follow': options['--follow'],
    'timestamps': options['--timestamps'],
}
```

seguins commented:
Thank you for the review. I applied fixes. I removed the part with threading and I rewrite the `test_logs_follow` test. I'm not sure it's the best way to test this case, but I haven't found better solution.

There is no change on `up` because it uses the `attach` api instead of `log`api.

dnephin commented:
> There is no change on up because it uses the attach api instead of log api.

`up` uses `LogPrinter`, which used to attach, but is now using `logs`, so I think it's no longer using attach.

seguins commented:
Indeed, I missed it ... Because docker-py has a backward compatibly, `up` still work. 

Should we set explicitly `follow=True` ?

dnephin commented:
I think it would be good to set that explicitly, yes

seguins commented:
`up` uses [a convergence_plan](https://github.com/docker/compose/blob/fc99c7ee1981a4dc2b125bfd08d4c31cbdb5435a/compose/project.py#L314). Convergence plan sets [`should_attach_logs = not detached`](https://github.com/docker/compose/blob/fc99c7ee1981a4dc2b125bfd08d4c31cbdb5435a/compose/service.py#L351) which has for effect to call `attach_log_stream` method of `container`. This method [attachs the container and stores the result in `log_stream`](https://github.com/docker/compose/blob/fc99c7ee1981a4dc2b125bfd08d4c31cbdb5435a/compose/container.py#L173).

`LogPrinter` checks if [`log_stream` is set](https://github.com/seguins/compose/blob/2227-improvements-logs/compose/cli/log_printer.py#L88) and in this case it's not call  the `logs` api. `up` will not call the `log` api but the `attach` api.

However, we need to set `follow=True` because it's used [for displaying the end of process](https://github.com/seguins/compose/blob/2227-improvements-logs/compose/cli/log_printer.py#L96).

Do you think there are enough tests if I just add to the `test_up_attached` test : `assert 'exited with code 0' in stdout` ?

dnephin commented:
I see, you're right it does use attach, I forgot about that. I think it's fine as is.

dnephin commented:
LGTM

shin- commented:
Might be a good idea to rerun the tests on janky? Code looks good overall.

aanand commented:
LGTM

JayHaoDing commented:
@aanand would you mind to comment if this merge will be done also to libcompose, which is used by rancher-compose? Thanks.

aanand commented:
@JayHaoDing It will have to be separately implemented. libcompose is not high priority at the moment - we have a Compose release very soon, so that's what we're focusing on.

IronicBadger commented:
I'm a little confused. This looks like it was merged but is the functionality available in compose in 1.7.1 ?
